iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/ml/ForecastOperator.java

```diff
@@ -0,0 +1,251 @@
+      int expectedPredictLength,
+      ExecutorService modelInferenceExecutor,
+      long maxRetainedSize,
+      long maxReturnSize) {
+    this.operatorContext = operatorContext;
+    this.child = child;
+    this.modelPath = modelPath;
+    this.inputTypeList = inputTypeList;
+    this.inputColumnNameList = inputColumnNameList;
+    this.expectedPredictLength = expectedPredictLength;
+    this.inputTsBlockBuilder = new TsBlockBuilder(inputTypeList);
+    this.modelInferenceExecutor = modelInferenceExecutor;
+    this.maxRetainedSize = maxRetainedSize;
+    this.maxReturnSize = maxReturnSize;
+  }
+
+  @Override
+  public OperatorContext getOperatorContext() {
+    return operatorContext;
+  }
+
+  @Override
+  public ListenableFuture<?> isBlocked() {
+    ListenableFuture<?> childBlocked = child.isBlocked();
+    boolean executionDone = forecastExecutionDone();
+    if (executionDone && childBlocked.isDone()) {
+      return NOT_BLOCKED;
+    } else if (childBlocked.isDone()) {
+      return forecastExecutionFuture;
+    } else if (executionDone) {
+      return childBlocked;
+    } else {
+      return successfulAsList(Arrays.asList(forecastExecutionFuture, childBlocked));
+    }
+  }
+
+  private boolean forecastExecutionDone() {
+    if (forecastExecutionFuture == null) {
+      return true;
+    }
+    return forecastExecutionFuture.isDone();
+  }
+
+  @Override
+  public boolean hasNext() throws Exception {
+    return !finished;
+  }
+
+  @Override
+  public TsBlock next() throws Exception {
+    if (forecastExecutionFuture == null) {
+      if (child.hasNextWithTimer()) {
+        TsBlock inputTsBlock = child.nextWithTimer();
+        if (inputTsBlock != null) {
+          appendTsBlockToBuilder(inputTsBlock);
+        }
+      } else {
+        submitForecastTask();
+      }
+      return null;
+    } else {
+      try {
+        if (!forecastExecutionFuture.isDone()) {
+          throw new IllegalStateException(
+              "The operator cannot continue until the forecast execution is done.");
+        }
+
+        TForecastResp forecastResp = forecastExecutionFuture.get();
+        if (forecastResp.getStatus().getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+          String message =
+              String.format(
+                  "Error occurred while executing forecast: %s",
+                  forecastResp.getStatus().getMessage());
+          throw new ModelInferenceProcessException(message);
+        }
+
+        finished = true;
+        TsBlock resultTsBlock =
+            new TsBlockSerde().deserialize(forecastResp.bufferForForecastResult());
+        resultTsBlock = modifyTimeColumn(resultTsBlock);
+        return resultTsBlock;
+      } catch (InterruptedException e) {
+        LOGGER.warn(
+            "{}: interrupted when processing write operation future with exception {}", this, e);
+        Thread.currentThread().interrupt();
+        throw new ModelInferenceProcessException(e.getMessage());
```

Comment:
Either log this exception and handle it, or rethrow it with some contextual information.

Reply:
done